### PR TITLE
New-TempUser Cmdlet

### PR DESCRIPTION
**Pull request by fprinvil-ms**
## Problem
When calling the REST API on a management VM like HCX authorized user credentials are needed to obtain an auth token and to make subsequent API calls.

## Solution
Author a cmdlet that provisions a temporary user with just enough access and returns the user credentials (username and password) to authenticate against vCenter.

## Testing
image

## Logs and metrics
n/a

## Upcoming changes
n/a

### CODE REVIEW

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 1]
```diff
@@ -1068,4 +1065,240 @@ function Set-ClusterDefaultStoragePolicy {
...
+            HelpMessage = 'User-Friendly name for the role')]
+        [ValidateNotNull()]
+        [string]
+        $userRole
+    )
+
+    if($privileges.Count -eq 0) { throw "If adding a temporary user, please ensure you pass in at least one privilege"}
+
+    $domain = "vsphere.local"
+    $group = "CloudAdmins"
```
Comment (et1975) [inline Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1111]: This group memembership probably should be optional - it has prohibitions that would prevent access to mgmt resources (like HCX appliance).

**Review by et1975 — CHANGES_REQUESTED** [review]
Let's move the private functions into a separate UserUtil.ps1 file, .-source the file in the psm1.

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 (hunk 1 above)
Comment (fprinvil-ms) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1]: Update the params to include the group of a non-mandatory parameter with the default value as "CloudAdmins"

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 2]
```diff
@@ -1068,4 +1065,240 @@ function Set-ClusterDefaultStoragePolicy {
...
+
+    .Parameter domain
+    Domain name (default value is vsphere.local)
+
+    .Example
+    Assert-UserExists -userName TempUser -domain "vsphere.local"
+#>
+Function Assert-UserExists {
+    [CmdletBinding()]
+    [AVSAttribute(30, UpdatesSDDC = $false)]
```
[the remark this answers is not in the document]
Comment (fprinvil-ms) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1]: Updated. Removed "AVSAttribute" from the cmdlet.

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 3]
```diff
@@ -1068,4 +1065,240 @@ function Set-ClusterDefaultStoragePolicy {
...
+
+    .Parameter userRole
+    Name of the role
+
+    .Example
+    Remove-TempUser -userName TempUser -userRole TempRole
+#>
+function Remove-TempUser {
+    [CmdletBinding(PositionalBinding = $false)]
+    [AVSAttribute(10, UpdatesSDDC = $false)]
```
[the remark this answers is not in the document]
Comment (fprinvil-ms) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1]: Updated. Removed "AVSAttribute" from the cmdlet

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 4]
```diff
@@ -1068,4 +1065,240 @@ function Set-ClusterDefaultStoragePolicy {
...
+#>
+Function Assert-UserExists {
+    [CmdletBinding()]
+    [AVSAttribute(30, UpdatesSDDC = $false)]
+    param(
+        [parameter(
+            Mandatory=$false,
+            HelpMessage = "User name")]
+        [string]
+        $userName = "TempUser",
```
[the remark this answers is not in the document]
Comment (fprinvil-ms) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1]: Updated. Removed the default value and set the param to mandatory

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 5]
```diff
@@ -1068,4 +1065,240 @@ function Set-ClusterDefaultStoragePolicy {
...
+        return $false;
+    }
+}
+
+<#
+    .Synopsis
+    Generates a password with at least 2 uppercase, 4 lowercase, 4 digits & 2 special character (!@#$%^&*())
+#>
+Function New-RandomPassword {
+    Write-Host "Starting $($MyInvocation.MyCommand)..."
```
[the remark this answers is not in the document]
Comment (fprinvil-ms) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1]: Good catch. Updated to remove this output statement.

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 6]
```diff
@@ -1068,4 +1065,240 @@ function Set-ClusterDefaultStoragePolicy {
...
+    .Example
+    Assert-RoleExists -userRole <role>
+#>
+Function Assert-RoleExists {
+    [CmdletBinding()]
+    [AVSAttribute(30, UpdatesSDDC = $false)]
+    param(
+        [parameter(Mandatory=$false,
+            HelpMessage = "Role name")]
+        [string]$userRole = "TempRole"
```
[the remark this answers is not in the document]
Comment (fprinvil-ms) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1]: Updated. Removed the default value and set the param to mandatory

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 7]
```diff
@@ -1068,4 +1068,121 @@ function Set-ClusterDefaultStoragePolicy {
...
+    $HcxPreferredVersion = '4.3.2'
+    $SlashCommonTreshold = '90%'
+    $HcxScaledtNumCpu = 8
+    $HcxScaledMemoryGb = 24
+    $Found = $false
+
+    Write-Host "Identifying HCX Vm"
+    $VmsList = Get-VM
+    foreach ($Vm in $VmsList) {
+        if($Vm.Name.Contains("HCX-MGR")) {
```
Comment (et1975) [inline Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1091]: `scripting` user won't find this VM because it's in the management pool.

Comment (jtrammell-msft) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1091]: Whats the work around, since the user has to set the CPU/Memory.

Comment (fprinvil-ms) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1]: Good catch. Updated to establish a connection to a vCenter Server system with temp user.

File: Microsoft.AVS.Management/Microsoft.AVS.Management.psm1 [hunk 8]
```diff
@@ -1068,4 +1068,131 @@ function Set-ClusterDefaultStoragePolicy {
...
+        $HcxServer = 'hcx'
+        $HcxPreferredVersion = '4.3.2'
+        $SlashCommonTreshold = '90%'
+        $HcxScaledtNumCpu = 8
+        $HcxScaledMemoryGb = 24
+        $Found = $false
+
+        Write-Host "Identifying HCX Vm"
+        $VmsList = Get-VM
+        foreach ($Vm in $VmsList) {
```
Comment (et1975) [inline Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1093]: Where is the connection with the new user?

Comment (fprinvil-ms) [reply Microsoft.AVS.Management/Microsoft.AVS.Management.psm1:1096]: Line 1084. Post the user creation.

**Pull request merged**